File: examples/litmus/cpp/lazy-cseq-2.0/modules/linearizability.c
// MAX_CLOCK should be the same as MAX_PUSHES to avoid the assume (*) in insert
#define MAX_PUSHES <insert-max_pushes-here>
#define MAX_ID_POP <insert-max_id_pop-here>
#define MAX_CLOCK <insert-max_clock-here>

typedef <insert-data_type> val_type;

unsigned char ids[MAX_PUSHES];
val_type values[MAX_PUSHES];
unsigned char finished[MAX_PUSHES];
unsigned char existence[MAX_PUSHES];
unsigned char time_at_call[MAX_PUSHES];
unsigned char time_at_return[MAX_PUSHES];

unsigned char clock = 0;
unsigned char saw_return = 0;

unsigned char pop_time_at_call[MAX_ID_POP];
_Bool saw_lp[MAX_ID_POP];

unsigned char nb_pushes_started_after_time[MAX_CLOCK];
unsigned char nb_pending_pushes_started_after_time[MAX_CLOCK];
unsigned char nb_pending_pushes_started_after_time_removed[MAX_CLOCK];

void __CSEQ_atomic_call_add_annotation(unsigned char id, val_type value) {
    unsigned char i;
    __CSEQ_assume ( 0 <= i & i < MAX_PUSHES );
    __CSEQ_assume ( existence[i] == 0 ); // & time_at_return[i] == 0
    // increments the clock if this call follows a return (i.e., clock counts the number of occurrences of return-call)
    clock += saw_return;
    saw_return = 0;
    //(*)__CSEQ_assume (clock < MAX_CLOCK);
    time_at_call[i] = clock;
    existence[i] = 1;
    ids[i] = id;
    values[i] = value;
    // from 0 to MAX_CLOCK-1; nb_pushes_started_after_time[i] = # pushes that started after time i
<update-nb_pushes-increment>

    // from 0 to MAX_CLOCK-1; nb_pending_pushes_started_after_time[i] = # pending pushes that started after time i
<update-nb_pending_pushes-increment>
}

void __CSEQ_atomic_return_add_annotation(unsigned char id) {
    // if the pushed value has been already removed by a pop, then there is nothing to be done
    if (<insert-id_check-here>)
        return;

    unsigned char i;
    __CSEQ_assume ( 0 <= i & i < MAX_PUSHES );
    __CSEQ_assume ( existence[i] == 1 & ids[i] == id );
    finished[i] = 1;
    time_at_return[i] = clock;
    saw_return = 1;

    // this is no longer a pending remove
<update-nb_pending_pushes-decrement>
}

void __CSEQ_atomic_remove_empty_annotation(unsigned char id) {
    // either all pushes have been removed or they are pending (and can be linearized after this pop returning empty)
    __CSEQ_assert(<insert-assert-empty>);

    // Mark linearization point
    saw_lp[id] = 1;
}

void __CSEQ_atomic_remove_annotation(unsigned char id, val_type value) {
    // there must exist a push adding this value to the stack
    __CSEQ_assert(<insert-assert-existence>);
    unsigned char i;
    __CSEQ_assume ( 0 <= i & i < MAX_PUSHES );
    __CSEQ_assume ( existence[i] == 1 & values[i] == value );
    /* either (1) the push is pending (thus the push can be linearized before)
              (2) there is no other finished push (which hasn't been removed) between the return of the removed push and the call of this pop
                        |——| |——| pop
                        0  0 1  1 1
              (3) there is no other *finished* push starting after this push. Since pending pushes may never be closed (by __CSEQ_atomic_return_push_annotation, because they are removed by a pop), we have to count those that are removed before being closed.
    */
    __CSEQ_assert( finished[i] == 0 | pop_time_at_call[id] <= time_at_return[i] |  nb_pushes_started_after_time[ time_at_return[i] + 1 ] == (nb_pending_pushes_started_after_time[ time_at_return[i] + 1 ] - nb_pending_pushes_started_after_time_removed[ time_at_return[i] + 1 ]) );  // nb_pushes_started_after_time[ time_at_return[i] + 1 ] == 0

    // from 0 to MAX_CLOCK; nb_pending_pushes_started_after_time[i] = # pending pushes that started after time i
<update-nb_pushes-decrement>

    if (finished[i] == 0) {
        // from 0 to MAX_OPERATIONS-1; nb_pending_pushes_started_after_time_removed[i] = # pending pushes that started at time i and have been removed
<update-nb_pending_pushes_remove-increment>
    }

    // reset the used PUSH index
    existence[i] = 0;
    finished[i] = 0;
    ids[i] = 0;
    values[i] = 0;
    time_at_call[i] = 0;
    time_at_return[i] = 0;

    // Mark linearization point
    saw_lp[id] = 1;
}

File: examples/litmus/cpp/lazy-cseq-2.0/modules/linearizability.py
""" CSeq C Sequentialization Framework
    linearizability module
    written by Truc Nguyen Lam, University of Southampton.
"""
VERSION = 'linearizability-2017.04.03' # initial version
"""
Description:
    This module instruments linearizability (forward simulation).
    For more details, please refer to https://arxiv.org/abs/1702.02705

    Parse string inside __CSEQ_rawline to identify linearization point
    Write to a header file linearizability instrumentation

    Overview of Linearizability module

    Sourcefile -> instrument ->

TODO:
    - Receive parameters (MAX_PUSHES, etc.)
    - Insert linearization point (manually or automatically)

Changelog:
    2017.04.03  1st version
"""
import core.common, core.module, core.utils


class linearizability(core.module.Translator):
    def init(self):
        self.addInputParam('li-max-add', 'maximum number of add (push or enqueue) action', 'p', '1', False)
        self.addInputParam('li-max-id-remove', 'maximum id of remove (pop or dequeue) actions', 'p', '1', False)
        self.addInputParam('li-max-clock', 'maximum clock tick', 'p', '1', False)
        self.addInputParam('li-data-type', 'data type (please wrap in \"\" if there is space in data type)', 'p', 'int', False)
        # TODO(me): add thread simulation
        # self.addInputParam('li-simulate', 'automatically simulate by configuration file', '', '', False)

    def loadfromstring(self, string, env):
        # backend = self.getInputParamValue('backend')
        MAX_PUSHES = int(self.getInputParamValue('li-max-add'))
        MAX_ID_POP = int(self.getInputParamValue('li-max-id-remove'))
        MAX_CLOCK = int(self.getInputParamValue('li-max-clock'))
        data_type = self.getInputParamValue('li-data-type')

        super(self.__class__, self).loadfromstring(string, env)

        header = core.utils.printFile("modules/linearizability.c")

        header = header.replace("<insert-max_pushes-here>", str(MAX_PUSHES))
        header = header.replace("<insert-max_id_pop-here>", str(MAX_ID_POP))
        header = header.replace("<insert-max_clock-here>", str(MAX_CLOCK))

        header = header.replace("<insert-data_type>", data_type)

        # inside function __VERIFIER_atomic_call_ADD_annotation
        nb_pushes_increment = ''
        for i in range(0, MAX_PUSHES):
            nb_pushes_increment += '    nb_pushes_started_after_time[%s] += (clock >= %s);\n' % (i, i)
        header = header.replace("<update-nb_pushes-increment>", nb_pushes_increment)

        nb_pending_pushes_increment = ''
        for i in range(0, MAX_PUSHES):
            nb_pending_pushes_increment += '    nb_pending_pushes_started_after_time[%s] += (clock >= %s);\n' % (i, i)
        header = header.replace("<update-nb_pending_pushes-increment>", nb_pending_pushes_increment)

        # inside function __VERIFIER_atomic_return_ADD_annotation
        header = header.replace("<insert-id_check-here>", ' & '.join(['ids[%s] != id' % i for i in range(0, MAX_PUSHES)]))

        nb_pending_pushes_decrement = ''
        for i in range(0, MAX_PUSHES):
            nb_pending_pushes_decrement += '    nb_pending_pushes_started_after_time[%s] -= ( time_at_call[i] >= %s);\n' % (i, i)
        header = header.replace("<update-nb_pending_pushes-decrement>", nb_pending_pushes_decrement)

        # inside function __VERIFIER_atomic_remove_empty_annotation
        assert_empty = ' & '.join(['(existence[%s] == 0 | finished[%s] == 0 | pop_time_at_call[id] <= time_at_return[%s])' % (i, i, i) for i in range(0, MAX_PUSHES)])
        header = header.replace("<insert-assert-empty>", assert_empty)

        # inside function __VERIFIER_atomic_remove_annotation
        assert_existence = ' | '.join(['(existence[%s] == 1 & values[%s] == value)' % (i, i) for i in range(0, MAX_PUSHES)])
        header = header.replace("<insert-assert-existence>", assert_existence)

        nb_pushes_decrement = ''
        for i in range(0, MAX_PUSHES):
            nb_pushes_decrement += '    nb_pushes_started_after_time[%s] -= ( time_at_call[i] >= %s);\n' % (i, i)
        header = header.replace("<update-nb_pushes-decrement>", nb_pushes_decrement)

        nb_pending_pushes_remove_increment = ''
        for i in range(0, MAX_PUSHES):
            nb_pending_pushes_remove_increment += '        nb_pending_pushes_started_after_time_removed[%s] += (time_at_call[i] >= %s);\n' % (i, i)
        header = header.replace("<update-nb_pending_pushes_remove-increment>", nb_pending_pushes_remove_increment)

        # Write to a header file, later get inserted by instrument module
        headerfilename = env.inputfile[:-2] + '_li.h'
        core.utils.saveFile(headerfilename, header)
        self.setOutputParam('liheader', headerfilename)

    def visit_FuncCall(self, n):
        # Looking for Linearization point, by __CSEQ_rawline
        fref = self._parenthesize_unless_simple(n.name)
        if fref == '__CSEQ_rawline':
            text = self.visit(n.args).strip('"')
            if text.startswith("@linear"):   # Only consider @linear at the moment
                text = text[8:]
                return text.replace("__VERIFIER_", "__CSEQ_")

        return fref + '(' + self.visit(n.args) + ')'
